File: java/tools/src/main/java/io/github/ajoz/workshop/fp/tools/Predicate.java
package io.github.ajoz.workshop.fp.tools;

@FunctionalInterface
public interface Predicate<A> {
    
    boolean test(final A value);

    // foo
    default Predicate<A> and(final Predicate<A> other) {
        return value -> this.test(value) && other.test(value);
    }

    // bar
    default Predicate<A> or(final Predicate<A> other) {
        return value -> this.test(value) || other.test(value);
    }

    default Predicate<A> not() {
        return value -> !this.test(value);
    }

    default Predicate<A> xor(final Predicate<A> other) {
        return value -> {
            final boolean p = this.test(value);
            final boolean q = other.test(value);
            return (p || q) && !(p && q);
        };
    }

    default Function1<A, Boolean> asFunction1() {
        return this::test;
    }
}

File: kotlin/tools/src/main/kotlin/org/artrev/workshop/fp/tools/Predicate.kt
package org.artrev.workshop.fp.tools

infix fun <A> ((A) -> Boolean).and(other: (A) -> Boolean): (A) -> Boolean =
        { value -> this(value) && other(value) }

infix fun <A> ((A) -> Boolean).or(other: (A) -> Boolean): (A) -> Boolean =
        { value -> this(value) || other(value) }

fun <A> ((A) -> Boolean).not(): (A) -> Boolean =
        { value -> !this(value) }

infix fun <A> ((A) -> Boolean).xor(other: (A) -> Boolean): (A) -> Boolean =
        { value ->
            val p = this(value)
            val q = other(value)
            (p || q) && !(p && q)
        }